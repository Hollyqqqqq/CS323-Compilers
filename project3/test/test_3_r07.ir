FUNCTION gcd1 :
PARAM v1
PARAM v2
t5 := v1 / v2
t4 := t5 * v2
v4 := v1 - t4
LABEL label1 :
IF v4 != #0 GOTO label2
GOTO label3
LABEL label2 :
v1 := v2
v2 := v4
t19 := v1 / v4
t18 := t19 * v4
v4 := v1 - t18
GOTO label1
LABEL label3 :
RETURN v2

FUNCTION gcd2 :
PARAM v5
PARAM v6
IF v5 > v6 GOTO label4
GOTO label5
LABEL label4 :
t33 := v5 / v6
t32 := t33 * v6
v7 := v5 - t32
IF v7 == #0 GOTO label7
GOTO label8
LABEL label7 :
RETURN v6
GOTO label5
LABEL label8 :
ARG v7
ARG v6
t40 := CALL gcd2
RETURN t40
LABEL label5 :
t45 := v6 / v5
t44 := t45 * v5
v7 := v6 - t44
IF v7 == #0 GOTO label10
GOTO label11
LABEL label10 :
RETURN v5
LABEL label11 :
ARG v7
ARG v5
t52 := CALL gcd2
RETURN t52

FUNCTION main :
READ v12
READ v13
ARG v13
ARG v12
t56 := CALL gcd1
WRITE t56
ARG v13
ARG v12
t58 := CALL gcd2
WRITE t58
RETURN #0

