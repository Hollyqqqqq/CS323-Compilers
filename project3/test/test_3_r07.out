Program (1)
  IncludeList (1)
  ExtDefList (1)
    ExtDef (1)
      Specifier (1)
        TYPE: int
      FunDec (1)
        ID: gcd1
        LP
        VarList (1)
          ParamDec (1)
            Specifier (1)
              TYPE: int
            VarDec (1)
              ID: a
          COMMA
          VarList (1)
            ParamDec (1)
              Specifier (1)
                TYPE: int
              VarDec (1)
                ID: b
        RP
      CompSt (2)
        LC
        DefList (3)
          Def (3)
            Specifier (3)
              TYPE: int
            DecList (3)
              Dec (3)
                VarDec (3)
                  ID: result
            SEMI
          DefList (4)
            Def (4)
              Specifier (4)
                TYPE: int
              DecList (4)
                Dec (4)
                  VarDec (4)
                    ID: temp
              SEMI
            DefList (1)
        StmtList (5)
          Stmt (5)
            Exp (5)
              Exp (5)
                ID: temp
              ASSIGN
              Exp (5)
                Exp (5)
                  ID: a
                MINUS
                Exp (5)
                  Exp (5)
                    Exp (5)
                      ID: a
                    DIV
                    Exp (5)
                      ID: b
                  MUL
                  Exp (5)
                    ID: b
            SEMI
          StmtList (6)
            Stmt (6)
              WHILE
              LP
              Exp (6)
                Exp (6)
                  ID: temp
                NE
                Exp (6)
                  INT: 0
              RP
              Stmt (6)
                CompSt (6)
                  LC
                  DefList (1)
                  StmtList (7)
                    Stmt (7)
                      Exp (7)
                        Exp (7)
                          ID: a
                        ASSIGN
                        Exp (7)
                          ID: b
                      SEMI
                    StmtList (8)
                      Stmt (8)
                        Exp (8)
                          Exp (8)
                            ID: b
                          ASSIGN
                          Exp (8)
                            ID: temp
                        SEMI
                      StmtList (9)
                        Stmt (9)
                          Exp (9)
                            Exp (9)
                              ID: temp
                            ASSIGN
                            Exp (9)
                              Exp (9)
                                ID: a
                              MINUS
                              Exp (9)
                                Exp (9)
                                  Exp (9)
                                    ID: a
                                  DIV
                                  Exp (9)
                                    ID: b
                                MUL
                                Exp (9)
                                  ID: b
                          SEMI
                        StmtList (1)
                  RC
            StmtList (11)
              Stmt (11)
                Exp (11)
                  Exp (11)
                    ID: result
                  ASSIGN
                  Exp (11)
                    ID: b
                SEMI
              StmtList (12)
                Stmt (12)
                  RETURN
                  Exp (12)
                    ID: result
                  SEMI
                StmtList (1)
        RC
    ExtDefList (14)
      ExtDef (14)
        Specifier (14)
          TYPE: int
        FunDec (14)
          ID: gcd2
          LP
          VarList (14)
            ParamDec (14)
              Specifier (14)
                TYPE: int
              VarDec (14)
                ID: x
            COMMA
            VarList (14)
              ParamDec (14)
                Specifier (14)
                  TYPE: int
                VarDec (14)
                  ID: y
          RP
        CompSt (15)
          LC
          DefList (16)
            Def (16)
              Specifier (16)
                TYPE: int
              DecList (16)
                Dec (16)
                  VarDec (16)
                    ID: remainder
                  ASSIGN
                  Exp (16)
                    INT: 0
              SEMI
            DefList (1)
          StmtList (17)
            Stmt (17)
              IF
              LP
              Exp (17)
                Exp (17)
                  ID: x
                GT
                Exp (17)
                  ID: y
              RP
              Stmt (17)
                CompSt (17)
                  LC
                  DefList (1)
                  StmtList (18)
                    Stmt (18)
                      Exp (18)
                        Exp (18)
                          ID: remainder
                        ASSIGN
                        Exp (18)
                          Exp (18)
                            ID: x
                          MINUS
                          Exp (18)
                            Exp (18)
                              Exp (18)
                                ID: x
                              DIV
                              Exp (18)
                                ID: y
                            MUL
                            Exp (18)
                              ID: y
                      SEMI
                    StmtList (19)
                      Stmt (19)
                        IF
                        LP
                        Exp (19)
                          Exp (19)
                            ID: remainder
                          EQ
                          Exp (19)
                            INT: 0
                        RP
                        Stmt (20)
                          RETURN
                          Exp (20)
                            ID: y
                          SEMI
                        ELSE
                        Stmt (22)
                          RETURN
                          Exp (22)
                            ID: gcd2
                            LP
                            Args (22)
                              Exp (22)
                                ID: y
                              COMMA
                              Args (22)
                                Exp (22)
                                  ID: remainder
                            RP
                          SEMI
                      StmtList (1)
                  RC
              ELSE
              Stmt (24)
                CompSt (24)
                  LC
                  DefList (1)
                  StmtList (25)
                    Stmt (25)
                      Exp (25)
                        Exp (25)
                          ID: remainder
                        ASSIGN
                        Exp (25)
                          Exp (25)
                            ID: y
                          MINUS
                          Exp (25)
                            Exp (25)
                              Exp (25)
                                ID: y
                              DIV
                              Exp (25)
                                ID: x
                            MUL
                            Exp (25)
                              ID: x
                      SEMI
                    StmtList (26)
                      Stmt (26)
                        IF
                        LP
                        Exp (26)
                          Exp (26)
                            ID: remainder
                          EQ
                          Exp (26)
                            INT: 0
                        RP
                        Stmt (27)
                          RETURN
                          Exp (27)
                            ID: x
                          SEMI
                        ELSE
                        Stmt (29)
                          RETURN
                          Exp (29)
                            ID: gcd2
                            LP
                            Args (29)
                              Exp (29)
                                ID: x
                              COMMA
                              Args (29)
                                Exp (29)
                                  ID: remainder
                            RP
                          SEMI
                      StmtList (1)
                  RC
            StmtList (1)
          RC
      ExtDefList (33)
        ExtDef (33)
          Specifier (33)
            TYPE: int
          FunDec (33)
            ID: main
            LP
            RP
          CompSt (34)
            LC
            DefList (35)
              Def (35)
                Specifier (35)
                  TYPE: int
                DecList (35)
                  Dec (35)
                    VarDec (35)
                      ID: m
                    ASSIGN
                    Exp (35)
                      READ
                      LP
                      RP
                SEMI
              DefList (36)
                Def (36)
                  Specifier (36)
                    TYPE: int
                  DecList (36)
                    Dec (36)
                      VarDec (36)
                        ID: n
                      ASSIGN
                      Exp (36)
                        READ
                        LP
                        RP
                  SEMI
                DefList (1)
            StmtList (37)
              Stmt (37)
                Exp (37)
                  WRITE
                  LP
                  Exp (37)
                    ID: gcd1
                    LP
                    Args (37)
                      Exp (37)
                        ID: m
                      COMMA
                      Args (37)
                        Exp (37)
                          ID: n
                    RP
                  RP
                SEMI
              StmtList (38)
                Stmt (38)
                  Exp (38)
                    WRITE
                    LP
                    Exp (38)
                      ID: gcd2
                      LP
                      Args (38)
                        Exp (38)
                          ID: m
                        COMMA
                        Args (38)
                          Exp (38)
                            ID: n
                      RP
                    RP
                  SEMI
                StmtList (39)
                  Stmt (39)
                    RETURN
                    Exp (39)
                      INT: 0
                    SEMI
                  StmtList (1)
            RC
        ExtDefList (1)
